pkg/workceptor/kubernetes.go

```diff
@@ -224,6 +234,105 @@ func podRunningAndReady(kw KubeUnit) func(event watch.Event) (bool, error) {
 	return inner
 }
 
+// podInfrastructureFailure checks if the pod has failed to start due to infrastructure issues.
+func (ku KubeAPIWrapper) podInfrastructureSuccess(pod *corev1.Pod) (bool, string, error) {
+	if pod == nil {
+		return false, "pod is nil", fmt.Errorf("pod is nil")
+	}
+
+	switch pod.Status.Phase {
+	case corev1.PodFailed:
+		for _, cs := range pod.Status.ContainerStatuses {
+			if cs.State.Terminated != nil {
+				switch cs.State.Terminated.ExitCode {
+				case 0:
+					break
+				default:
+					return false, cs.State.Terminated.Reason, nil
+				}
+			}
+		}
+		return true, "", nil
+
+	case corev1.PodSucceeded:
+		return true, "", nil
+
+	case corev1.PodRunning, corev1.PodPending:
+		return true, fmt.Sprintf("pod phase: %s", pod.Status.Phase), fmt.Errorf("pod not in terminal state")
+
+	default:
+		return false, fmt.Sprintf("unknown phase: %s", pod.Status.Phase), fmt.Errorf("invalid pod phase")
+	}
+}
+
+// PodApplicationSuccess checks if the pod has successfully completed its application logic.
+func (ku KubeAPIWrapper) podApplicationSuccess(pod *corev1.Pod) (bool, string, error) {
+	if pod == nil {
+		return false, "pod is nil", fmt.Errorf("pod is nil")
+	}
+
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.State.Terminated == nil {
+			return true, "container has not terminated", nil
+		}
+		if cs.State.Terminated.ExitCode != 0 {
+			return false, fmt.Sprintf("container %s exited with code %d: %s", cs.Name, cs.State.Terminated.ExitCode, cs.State.Terminated.Reason), nil
+		}
+	}
+
+	return true, "", nil
+}
+
+func (ku KubeAPIWrapper) WaitForPodCompleted(pod *corev1.Pod, clientset kubernetes.Interface, timeout time.Duration) (*corev1.Pod, error) {
+	namespace := pod.Namespace
+	name := pod.Name
+	interval := 1 * time.Second
+	latestPod := pod
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return latestPod, fmt.Errorf("timeout: pod did not complete within %s", timeout)
```

Comment:
```suggestion
			return latestPod, fmt.Errorf("timeout: pod %s/%s did not complete within %s", pod.Namespace, pod.Name, timeout)
```

We are going to need this multiple times. Would it make sense to make this a function that accepts pod and a string as arguments and it logs the error with the pod context?

Reply:
Cleaned this up a slightly different way. take a look please!

Reply:
Since we are not having a dedicated logging function, can we add the context of the pod name and pod namespace here?